packages/process/process_monitor.pony

```diff
@@ -295,20 +305,37 @@ actor ProcessMonitor
   fun _child(
     path: String,
     argp: Array[Pointer[U8] tag],
-    envp: Array[Pointer[U8] tag])
+    envp: Array[Pointer[U8] tag],
+    wdir: (FilePath | None))
   =>
     """
-    We are now in the child process. We redirect STDIN, STDOUT and STDERR
-    to their pipes and execute the command. The command is executed via
-    execve which does not return on success, and the text, data, bss, and
-    stack of the calling process are overwritten by that of the program
-    loaded. We've set the FD_CLOEXEC flag on all file descriptors to ensure
-    that they are all closed automatically once @execve gets called.
+    We are now in the child process. If a working directory was passed to
+    the child, a chdir is attempted. We redirect STDIN, STDOUT and STDERR to
+    their pipes and execute the command. The command is executed via execve which
+    does not return on success, and the text, data, bss, and stack of the
+    calling process are overwritten by that of the program loaded. We've set the
+    FD_CLOEXEC flag on all file descriptors to ensure that they are all closed
+    automatically once @execve gets called.
     """
     ifdef posix then
+      match wdir
+      | let d: FilePath =>
+        let dir: Pointer[U8] tag = d.path.cstring()
+
+        if 0 > @chdir[I32](dir) then
+          let errno: I32 = @pony_os_errno()
+          let strerror = @strerror[Pointer[U8]](errno)
+          let size = @strlen[USize](strerror)
+          let len = @write[ISize](_errpipe_write, strerror, size)
```

Comment:
Sorry I'm missing it, but who/where is the ultimate reader of this string that gets written?

Do we care about how it is delimited?

Reply:
The error str is captured in `_handle_event` and n `_pending_reads` it is passed to `ProcessNotify.stderr` and the return code `71` is returned from the child.

I would prefer to have [ProcessNotify.failed](https://stdlib.ponylang.io/process-ProcessNotify/#failed) being called with a new kind of Error based upon the `errno`. Maybe a class called `ChDirError` with an `I32` field called `errno` that contains the errno? Or maybe just a primitive.

Emitting the `strerror` on the child's `stderr` pipe might be misleading, as the caller might think, the `strerror` came from the actual process, although we didn't even executed it. I would vote for not emitting it there. Just send the `errno` through the pipe.

Reply:
Yeah, I would agree - let's not mix this into `stderr`.